### ISSUE TITLE
Type Variables and Sort Checking

### ISSUE THREAD

**Issue by ranjitjhala** [ucsd-progsys/liquid-fixpoint#98]
@nikivazou  see tests/todo/RG.hs.fq which is a shortened form of the .fq file obtained from: 

```
module RG where
data RGRef a
{-@ measure tv :: RGRef a -> a @-}
{-@ qualif TERMINALVALUE(r:RGRef a): (tv r) @-}
```

The enhancements to the polymorphic inference introduced  
in commit: 9608cb6e4dd33bf142b50db4630c629defceb91e 
named "another hack to fix sortchecking: refreshing type variables"
allow **unification** of the sort-variable for `a` with `bool` (or `prop`). 

This is a bug, because well, the `a` is "rigid" i.e. a monomorphic 
type at this point, and so should not be further unified with anything 
(let alone bool).

The code in that commit  lets liquid-fixpoint think that the qualifier 
`TERMINALVALUE` is well-sorted, which then **crashes** Z3 
because all values of sort `a` are embedded as `int` and not `bool`.

(Note that the parent commit seems fine -- I'm guessing its because 
the earlier "more conservative" polymorphic instantiation mechanism
rejected this qualifier is ill-formed out of hand.)

To reproduce;

```
$ fixpoint.native -v 1 -notruekvars -refinesort -nosimple -strictsortcheck -sortedquals -no-gen-qual-sorts -smtsolver z3 -out ./.liquid/RG.hs.fqout RG.hs.fq 
```

**Comment by nikivazou** [comment]
@ranjitjhala I believe that the problem here is that there is no checking that qualifiers are `bool` (or `prop`) values, neither at liquidHaskell not at fixpoint. Shouldn't we be checking that? 



For example, if you instead had `tv r` as a refinement, sort-check will fail:


```
constant tv : func(1, [RGRef @(0); @(0) ])

constraint: 
  env [r: {v: RGRef a | [] }]
  grd true
  lhs {v : L a | []}
  rhs {v : L a | [ ? (tv([r])) ]}
  id 1 
```

**Comment by ranjitjhala** [comment]
Don't follow -- the example below has no qualifiers?

On Thursday, September 10, 2015, Niki Vazou <notifications@github.com>
wrote:

> @ranjitjhala <https://github.com/ranjitjhala> I believe that the problem
> here is that there is no checking that qualifiers are bool (or prop)
> values, neither at liquidHaskell not at fixpoint. Shouldn't we be checking
> that?
>
> For example, if you instead had tv r as a refinement, sort-check will
> fail:
>
> constant tv : func(1, [RGRef @(0); @(0) ])
>
> constraint:
>   env [r: {v: RGRef a | [] }]
>   grd true
>   lhs {v : L a | []}
>   rhs {v : L a | [ ? (tv([r])) ]}
>   id 1
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/ucsd-progsys/liquid-fixpoint/issues/98#issuecomment-139418424>
> .
>


-- 
Ranjit.

**Comment by nikivazou** [comment]
It does not have qualifiers. But checking that `tv r :: bool` fails. 

If we were checking that the body of the qualifier is a `bool` then the initial example would create a proper error, instead of crashing at Z3. 

Is there a reason why we do not check well-formedness of qualifiers?

**Comment by ranjitjhala** [comment]
Well, the reason is just that the qualifiers are implicitly checked when
they are instantiated -- so they better be BOOL when you instantiate
them, or else they are not a valid instantiation. So I'm puzzled as to why
this should get past the instantiation. (This plus maybe some BAD reasons
due to our earlier hacky handling of polymorphism...)

On Thursday, September 10, 2015, Niki Vazou <notifications@github.com>
wrote:

> It does not have qualifiers. But checking that tv r :: bool fails.
>
> If we were checking that the body of the qualifier is a bool then the
> initial example would create a proper error, instead of crashing at Z3.
>
> Is there a reason why we do not check well-formedness of qualifiers?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/ucsd-progsys/liquid-fixpoint/issues/98#issuecomment-139419167>
> .
>


-- 
Ranjit.

**Comment by nikivazou** [comment]
Ok, I will check why it got through instantiation. Other than that, wouldn't it be better if we check well-formedness of qualifiers at liquidHaskell level, so that we provide better error messages there?

An error at instantiation will still happen at fixpoint, so we cannot provide good error messages.

**Comment by nikivazou** [comment]
Fix with #451

**Comment by ranjitjhala** [comment]
Thanks @nikivazou !!
